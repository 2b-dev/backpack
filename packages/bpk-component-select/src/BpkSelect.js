import React, { PropTypes } from 'react';

import './bpk-select.scss';

const BpkSelect = (props) => {
  const classNames = ['bpk-select'];
  const { valid, large, docked, dockedFirst, dockedMiddle, dockedLast, className, ...rest } = props;

  // Explicit check for false primitive value as undefined is
  // treated as neither valid nor invalid
  const isInvalid = valid === false;

  if (large) { classNames.push('bpk-select--large'); }
  if (docked) { classNames.push('bpk-select--docked'); }
  if (dockedFirst) { classNames.push('bpk-select--docked-first'); }
  if (dockedMiddle) { classNames.push('bpk-select--docked-middle'); }
  if (dockedLast) { classNames.push('bpk-select--docked-last'); }
  if (isInvalid) { classNames.push('bpk-select--invalid'); }
  if (className) { classNames.push(className); }

  return (
    <select className={classNames.join(' ')} aria-invalid={isInvalid} {...rest} />
  );
};

BpkSelect.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  className: PropTypes.string,
  valid: PropTypes.bool,
  large: PropTypes.bool,
  docked: PropTypes.bool,
  dockedFirst: PropTypes.bool,
  dockedMiddle: PropTypes.bool,
  dockedLast: PropTypes.bool,
};

BpkSelect.defaultProps = {
  className: null,
  valid: null,
  large: false,
  docked: false,
  dockedFirst: false,
  dockedMiddle: false,
  dockedLast: false,
};

export default BpkSelect;
