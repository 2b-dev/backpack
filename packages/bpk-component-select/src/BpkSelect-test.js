import React from 'react'
import renderer from 'react/lib/ReactTestRenderer'
import BpkSelect from './BpkSelect'

describe('BpkSelect', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <BpkSelect
        id='fruits'
        name='fruits'
        value='oranges'
        onChange={() => null}
        options={[
          { name: 'Apples', value: 'apples' },
          { name: 'Oranges', value: 'oranges' },
          { name: 'Pears', value: 'pears' },
          { name: 'Tomatos', value: 'tomatos', disabled: true }
        ]}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('should render correctly with a valid attribute equal to false', () => {
    const tree = renderer.create(
      <BpkSelect
        id='fruits'
        name='fruits'
        value='oranges'
        onChange={() => null}
        options={[
          { name: 'Apples', value: 'apples' },
          { name: 'Oranges', value: 'oranges' },
          { name: 'Pears', value: 'pears' },
          { name: 'Tomatos', value: 'tomatos', disabled: true }
        ]}
        valid={false}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with disabled attribute', () => {
    const tree = renderer.create(
      <BpkSelect
        id='fruits'
        name='fruits'
        value=''
        onChange={() => null}
        options={[
          { name: 'Apples', value: 'apples' },
          { name: 'Oranges', value: 'oranges' },
          { name: 'Pears', value: 'pears' },
          { name: 'Tomatos', value: 'tomatos', disabled: true }
        ]}
        disabled
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with "large" attribute', () => {
    const tree = renderer.create(
      <BpkSelect
        id='fruits'
        name='fruits'
        value=''
        onChange={() => null}
        options={[
          { name: 'Apples', value: 'apples' },
          { name: 'Oranges', value: 'oranges' },
          { name: 'Pears', value: 'pears' },
          { name: 'Tomatos', value: 'tomatos', disabled: true }
        ]}
        large
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly with "docked" attribute', () => {
    const tree = renderer.create(
      <BpkSelect
        id='fruits'
        name='fruits'
        value=''
        onChange={() => null}
        options={[
          { name: 'Apples', value: 'apples' },
          { name: 'Oranges', value: 'oranges' },
          { name: 'Pears', value: 'pears' },
          { name: 'Tomatos', value: 'tomatos', disabled: true }
        ]}
        docked
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

