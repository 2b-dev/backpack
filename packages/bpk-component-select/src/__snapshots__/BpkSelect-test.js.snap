exports[`BpkSelect should render correctly 1`] = `
<select
  className="bpk-select"
  disabled={false}
  id="fruits"
  name="fruits"
  onChange={[Function onChange]}
  value="oranges">
  <option
    disabled={undefined}
    hidden={undefined}
    value="apples">
    Apples
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="oranges">
    Oranges
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="pears">
    Pears
  </option>
  <option
    disabled={true}
    hidden={undefined}
    value="tomatos">
    Tomatos
  </option>
</select>
`;

exports[`BpkSelect should render correctly with "docked" attribute 1`] = `
<select
  className="bpk-select bpk-select--docked"
  disabled={false}
  id="fruits"
  name="fruits"
  onChange={[Function onChange]}
  value="">
  <option
    disabled={undefined}
    hidden={undefined}
    value="apples">
    Apples
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="oranges">
    Oranges
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="pears">
    Pears
  </option>
  <option
    disabled={true}
    hidden={undefined}
    value="tomatos">
    Tomatos
  </option>
</select>
`;

exports[`BpkSelect should render correctly with "large" attribute 1`] = `
<select
  className="bpk-select bpk-select--large"
  disabled={false}
  id="fruits"
  name="fruits"
  onChange={[Function onChange]}
  value="">
  <option
    disabled={undefined}
    hidden={undefined}
    value="apples">
    Apples
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="oranges">
    Oranges
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="pears">
    Pears
  </option>
  <option
    disabled={true}
    hidden={undefined}
    value="tomatos">
    Tomatos
  </option>
</select>
`;

exports[`BpkSelect should render correctly with a valid attribute equal to false 1`] = `
<select
  className="bpk-select bpk-select--invalid"
  disabled={false}
  id="fruits"
  name="fruits"
  onChange={[Function onChange]}
  value="oranges">
  <option
    disabled={undefined}
    hidden={undefined}
    value="apples">
    Apples
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="oranges">
    Oranges
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="pears">
    Pears
  </option>
  <option
    disabled={true}
    hidden={undefined}
    value="tomatos">
    Tomatos
  </option>
</select>
`;

exports[`BpkSelect should render correctly with disabled attribute 1`] = `
<select
  className="bpk-select"
  disabled={true}
  id="fruits"
  name="fruits"
  onChange={[Function onChange]}
  value="">
  <option
    disabled={undefined}
    hidden={undefined}
    value="apples">
    Apples
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="oranges">
    Oranges
  </option>
  <option
    disabled={undefined}
    hidden={undefined}
    value="pears">
    Pears
  </option>
  <option
    disabled={true}
    hidden={undefined}
    value="tomatos">
    Tomatos
  </option>
</select>
`;
